rosbag2_cpp/include/rosbag2_cpp/cache/circular_message_cache.hpp

```diff
@@ -0,0 +1,70 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_CPP__CACHE__CIRCULAR_MESSAGE_CACHE_HPP_
+#define ROSBAG2_CPP__CACHE__CIRCULAR_MESSAGE_CACHE_HPP_
+
+#include <memory>
+#include <mutex>
+#include <string>
+
+#include "rosbag2_cpp/cache/message_cache_circular_buffer.hpp"
+#include "rosbag2_cpp/visibility_control.hpp"
+
+#include "rosbag2_storage/serialized_bag_message.hpp"
+
+// This is necessary because of using stl types here. It is completely safe, because
+// a) the member is not accessible from the outside
+// b) there are no inline functions.
+#ifdef _WIN32
+# pragma warning(push)
+# pragma warning(disable:4251)
+#endif
+
+namespace rosbag2_cpp
+{
+namespace cache
+{
+
+class ROSBAG2_CPP_PUBLIC CircularMessageCache
+{
+public:
+  explicit CircularMessageCache(uint64_t max_buffer_size);
```

Comment:
This seems like it should inherit from a common interface as `MessageCache` - so that it can be passed to the same interfaces downstream without modification. I can see that right now `MessageCache` has the `swap` function built-in to logic we don't want for the snapshot version - but this could also be a configuration on the `MessagCache`. In that case maybe we could have a generic `MessageCache<BufferType>`, with one or two configuration variables, instead of a totally separate type that the consumers will have to understand?

Reply:
Nitpick. could be not addressed.
It's better to use `size_t` type for memory related variables to automatically accommodate for the underlying HW architecture.